[IMP] web_editor: enhance sub-options design within the editor

chgo-odoo opened the pull request:
This PR aims to enhance the sub-options within the editor, fixing an issue about sub-options being overemphasized.

task-4492278

----------

Prior to this PR, both options and sub-options were using the same color, making it a bit hard for the user to distinguish whether he was applying changes on an (sub)option or not.

There were also different issue like options that could be triggered by clicking on their label but without providing any visual feedback, nested sub-options in collapse not showing the user the all the options he unfolded in order to reach the one he's applying changes with.

This PR introduces different improvement to solve these issues by :

#### 1) Setting options to white, while sub-options will remain gray;

| Options and sub-options color - Master | This PR |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/db183e3f-dbdf-4bb2-9afb-6159fb156376) | ![image](https://github.com/user-attachments/assets/1d6fac00-01a2-4032-a425-a2b5d0b00278) | 


#### 2) Settings the unfold icon to white when hovering 
| Visual feedback when hovering a foldable icon - Master| This PR |
|--------|--------|
| <img width="109" alt="image" src="https://github.com/user-attachments/assets/0d2ee7ec-31e3-4295-a111-49df4bc46f8e" /> | <img width="109" alt="image" src="https://github.com/user-attachments/assets/d6dd51b6-1b95-4b78-8101-5b8c9bd83ef8" /> | 

#### 3) Setting a white color to the fold icon when unfolding nested options collapse, highlighting the path user is currently in.
| Options hierarchy visual feedback | This PR |
|--------|--------|
| <img width="290" alt="image" src="https://github.com/user-attachments/assets/67ee66db-8d5e-4ce5-bd95-ed919dd0b7d7" /> | <img width="289" alt="image" src="https://github.com/user-attachments/assets/a463fae7-64ac-433a-ae9e-0ae9bce548aa" /> | 

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/195781.png)](https://mergebot.odoo.com/odoo/odoo/pull/195781)

chgo-odoo commented:
Hello @stefanorigano 👋 

I added some extra changes to this, but everything we initially planned is there ! I also added screenshot for each change so you can grasp the overall modifications but it's way more visible on Runbot.

Let me know what you think 👍

stefanorigano reviewed (changes requested):
Hello @chgo-odoo , thanks 👍 
Here some remarks on the actual visual result, I didn't check the code just yet 😉

----

> 1) Setting options to white, while sub-options will remain gray;

ok

> 2) Settings the unfold icon to white when hovering a foldable option

It's not applied consistently, see "Theme" tab font-size
![image](https://github.com/user-attachments/assets/ea671757-7869-4330-a40b-cd898b5e8831).

> 3) Setting a white color on hover for sub-options that are fully clickable, providing a visual feedback to the user ;

Despite makes perfectly sense, I'm in doubt because we don't provide any visual feedback for the others ATM. Indeed we should define hover states for all the controllers, but it would be an huge task, probably out of scope of this PR. Eventually, once done globally, in this specific case the label should remain the same, while the switch should be highlighted instead. To think about that...

> 4) Setting a white color to the fold icon when unfolding nested options collapse, highlighting the path user is currently in.

A bit like point 3, but this is a very specific scenario so this change can stay 🙂.
Be aware that it's not applied consistently tho:
![image](https://github.com/user-attachments/assets/8695807f-5b11-4b39-9118-9f6122e85076)

---

I think we can improve even further by adjusting vertical spacing when sub-options are rendered. I did a fast test using the following code:

```scss
#oe_snippets > .o_we_customize_panel {
	.o_we_sublevel_1, o_we_sublevel_2 {
		margin-top: 4px;

		+ .o_we_sublevel_1, +.o_we_sublevel_2 {
			padding-bottom: 2px;
		}
	}
}
```
<img width="1220" alt="image" src="https://github.com/user-attachments/assets/3fb09158-6182-4a79-8afb-552b38e7b442" />

---

To investigate, not sure how huge such a change would be,  but if we could replace  the `"└"` pseudo-element with something like https://www.cssscript.com/tree-view-unlimited-nesting/ it would be great!

Thanks!

chgo-odoo commented:
Hello @stefanorigano 👋 

I adapted this according to your suggestions. About your last point and the sub-option icon, I gave it a try yesterday but it was heavy (~40 lines and it was not even handling all the cases). I do agree the current implementation is not perfect but I'd keep it as is, if we compare it to the required CSS to achieve something else.

lebl-odoo commented:
I love what it looks now, it is easier to read.

However, I'm not sure how the spacing between elements are computed, here it seems there is higher gap between Quality & Parallax than the others.
![image](https://github.com/user-attachments/assets/2715a3fb-5c5a-4826-9fb2-a58763399cb9)

On :hover, the caret change direction while it shouldn't. I had a look in master, it's not the case.
![image](https://github.com/user-attachments/assets/c3baca65-d281-42c0-934a-443fbda4b456)

Moreover, I'm wondering if the refactoring team should be aware of such a UI change, @sobo-odoo ?

chgo-odoo commented:
Thanks @lebl-odoo, should be fixed now 👍

stefanorigano reviewed (approved):
LGTM, thanks @chgo-odoo :)

Just a not-blocking nitpicking comment, but feel free to move to tech. if still confirmed by @lebl-odoo (I see that he reacted with a  "👍", but let's double-check).

When ready you should ping both the editor and website team.

Thanks again ! 


> About your last point and the sub-option icon, I gave it a try yesterday but it was heavy (~40 lines and it was not even handling all the cases). I do agree the current implementation is not perfect but I'd keep it as is, if we compare it to the required CSS to achieve something else.

Ok, I'll handle it in task-4573711 ;)

chgo-odoo commented:
Confirmed by @lebl-odoo 👍

bvr-odoo commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -279,7 +279,15 @@
+            color: $o-we-fg-light;
+
+            & + .o_we_sublevel_#{$level + 1} {
+                padding-bottom: 4px;
```
I suppose it comes from the suggestion of SRI, but it looks like it's not the expected result. It adds space in a place where it doesn't make sense, or maybe I'm missing something.
See here:
![image](https://github.com/user-attachments/assets/9efdb3c7-f747-4a4a-af0b-dc02344e52cb)

(Anyway, add a comment here to explain the purpose of this code, as it's not obvious.)

bvr-odoo commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -871,6 +879,8 @@
                 background-color: $o-we-sidebar-content-field-dropdown-bg;
                 box-shadow: $o-we-sidebar-content-field-dropdown-shadow;
             }
+
+
```
I wonder why you added these two lines.

bvr-odoo reviewed (changes requested):
A few small remarks 😉

chgo-odoo commented:
Hello @bvr-odoo, thanks for reviewing this one 👍 

I think we were actually overthinking the functional side, instead of using a padding on top of the margin, I rely on the default `8px` margin and apply a `4px` margin using the CSS variable **ONLY** if the sub-option is followed by another sub-option with the same level, or if an option is followed by a sub-option.

This simplifies drastically the implementation while keeping the idea we want to achieve.

Could you check if this works fine?

Thanks !

bso-odoo commented:
@robodoo delegate=bvr-odoo

bvr-odoo commented:
> Hello @bvr-odoo, thanks for reviewing this one 👍
> 
> I think we were actually overthinking the functional side, instead of using a padding on top of the margin, I rely on the default `8px` margin and apply a `4px` margin instead, using the CSS variable, **ONLY** if the sub-option is followed by another sub-option with the same level, or if an option is followed by a sub-option.
> 
> This simplifies drastically the implementation while keeping the idea we want to achieve.
> 
> Could you check if this works fine?
> 
> Thanks !

chgo-odoo commented:
Thanks @bvr-odoo 👍 

I just pushed another attempt but it's actually way trickier than expected... You can't really detect all those `d-none` elements as well as how much they are and then there is the issue of detecting the first element of the `we-customizeblock-option` which can have many different layout so it's quite easy to exclude a rule by mistake...

I'll check once the Runbot is built but there are still some corner cases with the current layout of the option and my implementation ... It might be nice to align the layout at some point.

I'll ping you once I think it's ready 👍

bvr-odoo reviewed (changes requested):
Hello @chgo-odoo ,
There are still some cases where the 8-pixel margin is applied when it shouldn't be. Can you check, please?

![image](https://github.com/user-attachments/assets/6fbc80f1-d51d-4727-88b4-6b6f775e0cb0)
![image](https://github.com/user-attachments/assets/1dddfcd1-7006-4419-ac7c-2fa6d6e01fb8)
![image](https://github.com/user-attachments/assets/5358f1ac-6015-4082-9cb1-622778564100)

chgo-odoo commented:
Hello @bvr-odoo 👋 

In the end, we decided to drop the spacing between sub-options as mentioned earlier.

It's nearly impossible to implement with the current layout and this is more subject to introduce unintended side effects rather than improving a behaviour.

bvr-odoo reviewed (approved):
LGTM, thanks for the changes

bvr-odoo commented:
@robodoo r+

chgo-odoo commented:
@bvr-odoo Sorry but I had to push force this one to rebase .. There was an issue syncing the PR I guess, it was flagged as blocked on mergebot while correctly built on runbot.

bvr-odoo commented:
@robodoo r+
